Add json configuration saving/loading

tbeswick96 opened the pull request:
**When merged this pull request will:**

- Removed saveable type definer
- Remove usage of vanilla saving mechanism
- Removes need for saving fix mod
- Uses Newtonsoft.Json for (de)serialisation
- Sorting configuration separated out to own class
- Sorting configurations saved and loaded at same time as vanilla saving and loading, asynchronously
- Only maintains latest 50 saves to avoid file/memory bloat
- Closes #38

tbeswick96 committed:
Add json configuration saving/loading

- Removed saveable type definer
- Uses Newtonsoft.Json for (de)serialisation
- Sorting configuration separated out to own class
- Sorting configurations saved and loaded at same time as vanilla saving and loading, asynchronously
- Only maintains latest 50 saves to avoid file/memory bloat

tbeswick96 committed:
Cleaned up jetbrains annotations

tbeswick96 committed:
Update readme with note about save fix mod

MLNW commented on Extension/Configuration/JsonFileService.cs:
```diff
@@ -0,0 +1,58 @@
+
+namespace YAPO.Configuration
+{
+    public static class JsonFileService
```
Since this only provides save and load methods for the sorter configuration it isn't really a general JsonFileService as the name implies

MLNW commented on Extension/Configuration/Models/SorterConfiguration.cs:
```diff
@@ -0,0 +1,18 @@
+    [UsedImplicitly]
+    public class SorterConfiguration
+    {
+        public SortDirection SortDirection;
+
+        public bool UpgradableOnTop;
+
+        public bool SortOrderOpposite;
+
+        public SortMode CurrentSortByMode;
+
+        public SortMode CurrentThenByMode;
```
Could get rid of the new lines in between to use the same style as other data classes in the code base

MLNW commented on Extension/Configuration/Models/SorterConfigurationSave.cs:
```diff
@@ -0,0 +1,14 @@
+    [UsedImplicitly]
+    public class SorterConfigurationSave
+    {
+        public string Id;
+        public DateTime LastSaved;
+        public SorterConfiguration Party;
+        public SorterConfiguration Other;
```
So the file itself does not store which save it is related to? May make sense to also store that

tbeswick96 replied:
Yeah, `Id` is the save name, the file stores multiple saves in one file

MLNW replied:
Then it may make sense to name it accordingly? This way someone needs to know that the ID is the save name for it to make sense.

tbeswick96 replied:
Yep good idea

MLNW commented on Extension/Configuration/Patches/MBSaveLoadPatches.cs:
```diff
@@ -0,0 +1,46 @@
+        {
+            public static void Postfix(string saveName)
+            {
+                States.CurrentSaveId = saveName.Replace(".tmp", "");
```
I'm sure this needs to be done but I don't know why just looking at this part of the code. A comment may actually be beneficial

tbeswick96 replied:
Good idea

MLNW commented on Extension/Services/TroopSorterService.cs:
```diff
@@ -1,47 +1,42 @@
-            heroTroops = CurrentThenByMode == SortMode.NONE || CurrentThenByMode == CurrentSortByMode ? heroTroops.SortBy(CurrentSortByMode, SortDirection).ToList() : heroTroops.SortBy(CurrentSortByMode, SortDirection).ThenSortBy(CurrentThenByMode, SortOrderOpposite ? SortDirection.InvertDirection() : SortDirection).ToList();
-
-            if (!UpgradableOnTop) return;
+            sortedTroops =
+                configuration.CurrentThenByMode == SortMode.NONE ||
+                configuration.CurrentThenByMode == configuration.CurrentSortByMode
+                    ? sortedTroops.SortBy(configuration.CurrentSortByMode, configuration.SortDirection).ToList()
+                    : sortedTroops.SortBy(configuration.CurrentSortByMode, configuration.SortDirection)
+                                  .ThenSortBy(configuration.CurrentThenByMode,
+                                              configuration.SortOrderOpposite
+                                                  ? configuration.SortDirection.InvertDirection()
+                                                  : configuration.SortDirection)
+                                  .ToList();
+            heroTroops =
+                configuration.CurrentThenByMode == SortMode.NONE ||
+                configuration.CurrentThenByMode == configuration.CurrentSortByMode
+                    ? heroTroops.SortBy(configuration.CurrentSortByMode, configuration.SortDirection).ToList()
+                    : heroTroops.SortBy(configuration.CurrentSortByMode, configuration.SortDirection)
+                                .ThenSortBy(configuration.CurrentThenByMode,
+                                            configuration.SortOrderOpposite
+                                                ? configuration.SortDirection.InvertDirection()
+                                                : configuration.SortDirection)
+                                .ToList();
```
I like the formatting change! A lot easier to see what is going on

MLNW commented on README.md:
```diff
@@ -11,7 +11,7 @@
 |------------|:---------:|-------|
 | [ModLib](https://www.nexusmods.com/mountandblade2bannerlord/mods/592) | ✓ | Used to provide settings and other goodies |
-| [Save Missing Module Fix](https://www.nexusmods.com/mountandblade2bannerlord/mods/282) | ✗ | Recommended as it ensures saves can be loaded when you unload this mod having saved a game with it enabled |
+| [ALTERNATIVE Save Missing Module Fix](https://github.com/bmountney/Aragas.AltSaveSystemFix/releases/tag/1.3.0.0) | ✗ | Recommended as it ensures saves can be loaded when you unload this mod having saved a game with it enabled. Tested as working with this mod (Note: Original version does not work correctly) |
```
As far as I understood the change this mod shouldn't be listed here at all anymore. Only reason would be for older versions but then it should be stated more clearly.

MLNW reviewed (changes requested):
A few smaller things. May not have been a thorough review due to the time of day.

tbeswick96 committed:
Pull request review changes

- Renamed JsonFileService to be more specific to its function
- Added comment about save name .tmp removal
- Removed whitespace
- Update readme with better not about save missing module fix mod

tbeswick96 committed:
Swapped lines in readme

tbeswick96 committed:
Rename sorter save ID to SaveName
